staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go

```diff
@@ -207,12 +208,24 @@ func shutdownInCloudProvider(ctx context.Context, cloud cloudprovider.Interface,
 		return false, errors.New("cloud provider does not support instances")
 	}
 
-	shutdown, err := instances.InstanceShutdownByProviderID(ctx, node.Spec.ProviderID)
-	if err == cloudprovider.NotImplemented {
-		return false, nil
+	providerID := node.Spec.ProviderID
+	if providerID == "" {
```

Comment:
having additional check is fine but for my understanding may i know why the node doesn't have provider id ? Is it because the node is registered manually? Because kubelet seems to make sure provider id is set https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_node_status.go#L388-L393

Reply:
I am by no means close to explain why that happens, but I dumped the whole Node object passed around and I can only confirm that for some reason it does not have provider id in the spec

```
&Node{ObjectMeta:{ci-ln-1fllgz2-c1627-c6gh5-master-0    178d7133-fb6d-4550-a575-f0bcbc0f430d 4505 0 2023-09-14 12:05:36 +0000 UTC <nil> <nil> map[beta.kubernetes.io/arch:amd64 beta.kubernetes.io/os:linux kubernetes.io/arch:amd64 kubernetes.io/hostname:ci-ln-1fllgz2-c1627-c6gh5-master-0 kubernetes.io/os:linux node-role.kubernetes.io/control-plane: node-role.kubernetes.io/master: node.openshift.io/os_id:rhcos] map[alpha.kubernetes.io/provided-node-ip:192.168.130.10,fd65:a1a8:60ad:271c::5a volumes.kubernetes.io/controller-managed-attach-detach:XXXX] [] [] []},Spec:NodeSpec{PodCIDR:,DoNotUseExternalID:,ProviderID:,Unschedulable:false,Taints:[]Taint{Taint{Key:node-role.kubernetes.io/master,Value:,Effect:NoSchedule,TimeAdded:<nil>,},Taint{Key:node.cloudprovider.kubernetes.io/uninitialized,Value:XXXX,Effect:NoSchedule,TimeAdded:<nil>,},Taint{Key:node.kubernetes.io/not-ready,Value:,Effect:NoSchedule,TimeAdded:<nil>,},},ConfigSource:nil,PodCIDRs:[],},Status:NodeStatus{Capacity:ResourceList{cpu: {{4 0} {<nil>} 4 DecimalSI},ephemeral-storage: {{128248430592 0} {<nil>}  BinarySI},hugepages-1Gi: {{0 0} {<nil>} 0 DecimalSI},hugepages-2Mi: {{0 0} {<nil>} 0 DecimalSI},memory: {{16759611392 0} {<nil>}  BinarySI},pods: {{250 0} {<nil>} 250 DecimalSI},},Allocatable:ResourceList{cpu: {{3500 -3} {<nil>} 3500m DecimalSI},ephemeral-storage: {{114349845518 0} {<nil>} 114349845518 DecimalSI},hugepages-1Gi: {{0 0} {<nil>} 0 DecimalSI},hugepages-2Mi: {{0 0} {<nil>} 0 DecimalSI},memory: {{15581011968 0} {<nil>}  BinarySI},pods: {{250 0} {<nil>} 250 DecimalSI},},Phase:,Conditions:[]NodeCondition{NodeCondition{Type:MemoryPressure,Status:False,LastHeartbeatTime:2023-09-14 12:11:02 +0000 UTC,LastTransitionTime:2023-09-14 12:05:36 +0000 UTC,Reason:KubeletHasSufficientMemory,Message:kubelet has sufficient memory available,},NodeCondition{Type:DiskPressure,Status:False,LastHeartbeatTime:2023-09-14 12:11:02 +0000 UTC,LastTransitionTime:2023-09-14 12:05:36 +0000 UTC,Reason:KubeletHasNoDiskPressure,Message:kubelet has no disk pressure,},NodeCondition{Type:PIDPressure,Status:False,LastHeartbeatTime:2023-09-14 12:11:02 +0000 UTC,LastTransitionTime:2023-09-14 12:05:36 +0000 UTC,Reason:KubeletHasSufficientPID,Message:kubelet has sufficient PID available,},NodeCondition{Type:Ready,Status:False,LastHeartbeatTime:2023-09-14 12:11:02 +0000 UTC,LastTransitionTime:2023-09-14 12:05:36 +0000 UTC,Reason:KubeletNotReady,Message:container runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady message:Network plugin returns error: No CNI configuration file in /etc/kubernetes/cni/net.d/. Has your network provider started?,},},Addresses:[]NodeAddress{NodeAddress{Type:InternalIP,Address:192.168.130.10,},NodeAddress{Type:InternalIP,Address:fd65:a1a8:60ad:271c::5a,},NodeAddress{Type:Hostname,Address:ci-ln-1fllgz2-c1627-c6gh5-master-0,},},DaemonEndpoints:NodeDaemonEndpoints{KubeletEndpoint:DaemonEndpoint{Port:10250,},},NodeInfo:NodeSystemInfo{MachineID:3c688e98dbd040e8828076dcd2aab565,SystemUUID:94481b42-1446-7993-f130-543b246a5a97,BootID:af3f4814-10a0-4efa-979e-610779962d55,KernelVersion:5.14.0-284.32.1.el9_2.x86_64,OSImage:Red Hat Enterprise Linux CoreOS 414.92.202309121329-0 (Plow),ContainerRuntimeVersion:cri-o://1.27.1-7.rhaos4.14.git57d0fe5.el9,KubeletVersion:v1.27.4+2c287eb,KubeProxyVersion:v1.27.4+2c287eb,OperatingSystem:linux,Architecture:amd64,},Images:[]ContainerImage{ContainerImage{Names:[registry.apps.build01-us-west-2.vmc.ci.openshift.org/ci-ln-1fllgz2/stable@sha256:f3d1f991c21485c0ce342c15e3364fbd65feaf8c8551dbc0472a187e0711e578],SizeBytes:801282455,},ContainerImage{Names:[registry.apps.build01-us-west-2.vmc.ci.openshift.org/ci-ln-1fllgz2/stable@sha256:23012b3380ffce706aa8f204cdc26745d8a69b0218150ec3bcb495202694fdab],SizeBytes:714641214,},ContainerImage{Names:[registry.apps.build01-us-west-2.vmc.ci.openshift.org/ci-ln-1fllgz2/stable@sha256:15f1974647ac40b8fc895474cc6db94a553929020a2492732b49e50e89830d74],SizeBytes:498685551,},ContainerImage{Names:[registry.apps.build01-us-west-2.vmc.ci.openshift.org/ci-ln-1fllgz2/stable@sha256:0720ec7a4f788ef0327c52d0c2ec732b34f4bb6bb68968d253761ff74394041f],SizeBytes:492764068,},ContainerImage{Names:[registry.apps.build01-us-west-2.vmc.ci.openshift.org/ci-ln-1fllgz2/stable@sha256:257eecb1b8a9bfd55fdd8c7c4044398de91241920088f7729fb3303241177e82],SizeBytes:477792737,},ContainerImage{Names:[registry.apps.build01-us-west-2.vmc.ci.openshift.org/ci-ln-1fllgz2/stable@sha256:0b996a8dd7422387944eba766203e40f519af120841e1bc5fa533f97c7bd7180],SizeBytes:405326430,},},VolumesInUse:[],VolumesAttached:[]AttachedVolume{},Config:nil,},}
```

Reply:
Understood. The reason i am asking is to see if there is another issue because of which we are not setting the provider id in the node. Can you check the kubelet config and logs?

Reply:
I doubt this is really the case in my environment but from just going over the code I can see that it can happen that we have cloud provider that doesn't use provider ids and only relies on node name, i.e.

```
providerID, err := cloudprovider.GetInstanceProviderID(ctx, cnc.cloud, types.NodeName(node.Name))
if err == cloudprovider.NotImplemented {
	// if the cloud provider being used does not support provider IDs,
	// we can safely continue since we will attempt to set node
	// addresses given the node name in getNodeAddressesByProviderIDOrName
	klog.Warningf("cloud provider does not set node provider ID, using node name to discover node %s", node.Name)
	return "", nil
}
```

in https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go#L580-L603

Reply:
got it. If you are using a provider which is open source can you share it. Would like to take a look once.